hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/SubstringFunction.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.string;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.TriExpression;
+import com.hazelcast.sql.impl.expression.math.MathFunctionUtils;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+public class SubstringFunction extends TriExpression<String> implements IdentifiedDataSerializable {
+    public SubstringFunction() {
+        // No-op
+    }
+
+    private SubstringFunction(Expression<?> input, Expression<?> start, Expression<?> length) {
+        super(input, start, length);
+    }
+
+    public static SubstringFunction create(Expression<?> input, Expression<?> start, Expression<?> length) {
+        return new SubstringFunction(input, start, length);
+    }
+
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity"})
+    @Override
+    public String eval(Row row, ExpressionEvalContext context) {
+        // Get input
+        String input;
+
+        try {
+            input = StringExpressionUtils.asVarchar(operand1, row, context);
+        } catch (Exception e) {
+            // Conversion to String failed. E.g. NPE on UserClass.toString()
+            throw QueryException.dataException("Failed to get value of input operand of SUBSTRING function", e);
```

Comment:
Maybe we should include the original exception in the message. The user won't see the cause, i wonder if it will even be in the member logs.

Also I couldn't find a test for this scenario.

"NPE on" -> "NPE in"

Reply:
This is a defensive handler "just in case". We are going to make expressions more restrictive, not allowing any argument here except for strings, so exceptions will never happen. Meanwhile, this handler is added just for some unexpected situations that I cannot even imagine how to reproduce.

Let's assume that this handler will be removed soon during expressions polishing.

Reply:
It might be better to catch the exc in `ObjectConverter.asVarchar()`, when it calls the `toString`. Then it would not be necessary to handle it specially here if I'm not mistaken.

To test it we can query an OBJECT value with broken toString method.

Reply:
`OBJECT` Is restricted on the upper level at the moment. We cannot pass a custom object here at the moment.